clippy_lints/src/indexing_slicing.rs

```diff
@@ -103,7 +104,15 @@ impl<'tcx> LateLintPass<'tcx> for IndexingSlicing {
             return;
         }
 
-        if let ExprKind::Index(array, index, _) = &expr.kind {
+        if let ExprKind::Index(array, index, _) = &expr.kind
+            && let expr_ty = cx.typeck_results().expr_ty(array)
+            && let mut deref = deref_chain(cx, expr_ty)
+            && deref.any(|l| {
+                l.peel_refs().is_slice()
+                    || l.peel_refs().is_array()
+                    || adt_has_inherent_method(cx, l.peel_refs(), sym!(get))
```

Comment:
Would it make sense to also check the result type of the `get` method here? I could easily build a type with Index returning a slice of bytes while `get` returns a usize.

Please add such a test case, too.